pkg/stash/with_azureblob.go

```diff
@@ -0,0 +1,156 @@
+		return nil, errors.E(op, err)
+	}
+	cred, err := azblob.NewSharedKeyCredential(conf.AccountName, conf.AccountKey)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	pipe := azblob.NewPipeline(cred, azblob.PipelineOptions{})
+	serviceURL := azblob.NewServiceURL(*accountURL, pipe)
+
+	containerURL := serviceURL.NewContainerURL(conf.ContainerName)
+
+	return func(s Stasher) Stasher {
+		return &azblobLock{containerURL, s, checker}
+	}, nil
+}
+
+type azblobLock struct {
+	containerURL azblob.ContainerURL
+	stasher      Stasher
+	checker      storage.Checker
+}
+
+type stashRes struct {
+	v   string
+	err error
+}
+
+func (s *azblobLock) Stash(ctx context.Context, mod, ver string) (newVer string, err error) {
+	const op errors.Op = "azblobLock.Stash"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+	leaseBlobName := "lease/" + config.FmtModVer(mod, ver)
+	leaseBlobURL := s.containerURL.NewBlockBlobURL(leaseBlobName)
+
+	leaseID, err := s.acquireLease(ctx, leaseBlobURL)
+	if err != nil {
+		return ver, errors.E(op, err)
+	}
+	defer func() {
+		const op errors.Op = "azblobLock.Unlock"
+		relErr := s.releaseLease(ctx, leaseBlobURL, leaseID)
+		if err == nil && relErr != nil {
+			err = errors.E(op, relErr)
+		}
+	}()
+	ok, err := s.checker.Exists(ctx, mod, ver)
+	if err != nil {
+		return ver, errors.E(op, err)
+	}
+	if ok {
+		return ver, nil
+	}
+	sChan := make(chan stashRes)
+	go func() {
+		v, err := s.stasher.Stash(ctx, mod, ver)
+		sChan <- stashRes{v, err}
+	}()
+
+	for {
+		select {
+		case sr := <-sChan:
+			if sr.err != nil {
+				err = errors.E(op, sr.err)
+				return ver, err
+			}
+			newVer = sr.v
+			return newVer, nil
+		case <-time.After(10 * time.Second):
+			err := s.renewLease(ctx, leaseBlobURL, leaseID)
+			if err != nil {
+				return ver, errors.E(op, err)
+			}
+		}
+	}
+}
+
+func (s *azblobLock) releaseLease(ctx context.Context, blobURL azblob.BlockBlobURL, leaseID string) error {
+	const op errors.Op = "azblobLock.releaseLease"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+	_, err := blobURL.ReleaseLease(ctx, leaseID, azblob.ModifiedAccessConditions{})
+	return err
+}
+
+func (s *azblobLock) renewLease(ctx context.Context, blobURL azblob.BlockBlobURL, leaseID string) error {
+	const op errors.Op = "azblobLock.renewLease"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+	_, err := blobURL.RenewLease(ctx, leaseID, azblob.ModifiedAccessConditions{})
+	return err
+}
+
+func (s *azblobLock) acquireLease(ctx context.Context, blobURL azblob.BlockBlobURL) (string, error) {
+	const op errors.Op = "azblobLock.acquireLease"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+	tctx, cancel := context.WithTimeout(ctx, 300*time.Second)
+	defer cancel()
+
+	// first we need to create a blob which can be then leased
+	_, err := blobURL.Upload(tctx, bytes.NewReader([]byte{1}), azblob.BlobHTTPHeaders{}, nil, azblob.BlobAccessConditions{})
```

Comment:
So it seems that the lease acquiring process has a timeout of 5 minutes, but the lease itself doesn't have a timeout. In other words, if the stashing process takes too long, the lease will keep getting refreshed every 10 seconds forever right? 

I know that our _main_ stasher has its own timeout of 10 minutes so just double checking that I'm following the logic here.

Reply:
@marwan-at-work 
Good question! I figured I will rely on the inner stasher timeout. Probably not a good idea if for some reason the stasher doesn't have a timeout and hangs forever.
For some reason I was convinced that the stasher timeout is 5 and not 10 min.
Which leads me to another question - if the stasher has timeout = 10 min shouldn't the lease acquiring process also have the same timeout? It will need to wait for the first request so max 10 min. I'm asking b/c the redis lock uses 5 min.

Reply:
@marpio yeah that's a good point. We should probably make the timeouts more consistent. I don't know if 5 is better than 10 but either way is cool with me. The whole point is that we have some sane timeout on the stashing process...I still think we don't need to make it configurable (just yet), so maybe the longer the better.